src/server.rs

```diff
@@ -1,62 +1,119 @@
-use futures::{future, executor, Async, Future, Stream};
+use bytes::Bytes;
+use futures::{Async, Poll, Future, Stream};
 use flate2::Compression;
-use flate2::read::GzEncoder;
+use flate2::write::GzEncoder;
 use hyper::{self, Body};
-// use hyper::header;
 use hyper::server::Service;
 use http::{self, header};
-use tokio_io::{AsyncRead, AsyncWrite};
+use tokio_io::AsyncRead;
 
-// use ::MaybeCompressed;
+use std::io::{Write, Cursor};
 
-use std::io::{Read, Write, Cursor};
-use std::thread;
+#[derive(Clone, Copy, Debug, Default)]
+pub struct Builder {
+    chunk_size: Option<usize>,
+    compression: Option<Compression>,
+}
 
-pub struct Gzip<T> {
+#[derive(Debug, Clone)]
+pub struct GzipChunked<T> {
     inner: T,
+    chunk_size: usize,
+    compression: Compression,
 }
 
-// fn map_response<B, F>(rsp: http::Response<B>, f: F)
-//     -> http::Response<MaybeCompressed<B>>
-// where
-//     F: Fn(B) -> MaybeCompressed<B>,
-//     B: Write,
-// {
-//     let status = rsp.status();
-//     let headers = rsp.headers().clone(); // TODO: remove clone.
-//     let body: B = *(rsp.body_ref().unwrap().clone());
-//     let body = f(body);
-//     http::Response::new()
-//         .with_status(status)
-//         .with_headers(headers)
-//         .with_body(body)
-// }
-
-impl<T> Gzip<T> {
+const MAX_CHUNK_SIZE: usize = 2048;
+
+struct ChunkingStream<B: AsyncRead> {
```

Comment:
I admit, I don't really understand what this struct is for. It's an adapter for `AsyncRead` into `Stream`?